### DISCUSSION TITLE
How to verify a mean/max grib2 variable?

### DISCUSSION THREAD

**Discussion by YaliMao-NOAA** [dtcenter/METplus#1915] [Configuration] [answered]
May I know how to differentiate and verify a mean/max grib2 variable of template 4.15 with the same parameter and category number, but with different types of statistical processing?

Here is grib2 doc link of template 4.15:
https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_temp4-15.shtml
And here is the type of statistical processing:
https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-10.shtml

I uploaded a sample data, WAFS_blended_2022110300f18.grib2 at https://ftp.emc.ncep.noaa.gov/unaff/ymao/metplus/
I want to verify mean/max ICIP (icing potential) of template 4.15, which has the same cat=19 parm=20. The only way to differentiate mean/max ICIP is to lookup the type of statistical processing: 0 - average/mean 2- max

MET_GRIB_TABLES doesn't work in this case, as described on METplus page:
https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#met-grib-tables

Thank you for the help!

**Comment by YaliMao-NOAA** [comment]
I took a look at the log file. Template 4.15 is not supported by METplus. I will figure out a way to walk around

**Comment by YaliMao-NOAA** [comment, marked as answer]
I moved on. I separated mean/max ICIP into two files, renamed max ICIP to another name, then combined them back together, then converted to template 4.0. But I needed to redefine these two fields in my own MET_GRIB_TABLES

**Reply by JohnHalleyGotway** [reply]
1. Yali, FYI, I'm on support desk today and did take a closer look at the data you provided. I always use the plot_data_plane tool to address this sort of thing.

Based on the output from wgrib2 where I see a variable named ICIP, I tried running:
```
plot_data_plane WAFS_blended_2022110300f18.grib2 ICIP_max.ps 'name="ICIP"; level="P500";' -v 3
```
And I see this warning message in the log output:
```
WARNING: MetGrib2DataFile::data_plane() -> Using the first of 2 matching records for "ICIP/P500":
WARNING: record 22 field 1: ipdtmpl[18] = 19, 20, 2, 0, 96, 0, 0, 1, 18, 100, 0, 50000, 255, 0, 0, 0, 3, 1
WARNING: record 23 field 1: ipdtmpl[18] = 19, 20, 2, 0, 96, 0, 0, 1, 18, 100, 0, 50000, 255, 0, 0, 2, 3, 1
WARNING: Try setting the 0-based "GRIB2_ipdtmpl_index" and "GRIB2_ipdtmpl_val" config file options to select one.
```
So MET (as you know) is not distinguishing between records 22 and 23. There are several ways you can distinguish between records in GRIB files and you can see those by searching for "GRIB2_" in this [section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=grib2_#fcst) of the MET User's Guide. In particular, I was hopeful that setting `GRIB2_stat_type = 2;` to filter based on the GRIB2 Table 4.10 would distinguish between them:
```
plot_data_plane WAFS_blended_2022110300f18.grib2 ICIP_max.ps 'name="ICIP"; level="P500"; GRIB2_stat_type=2; ' -v 3
```
But unfortunately it did not:
```
WARNING: MetGrib2DataFile::data_plane() -> No matching record found for 'ICIP/P500'
```
Take a look back at the original warning message though. We provided the `GRIB2_ipdtmple_index` and `GRIB2_ipdtmpl_val` options as a *catch-all*. So if all else fails, you can always use those to distinguish between multiple matching records. Looking carefully at the warning message, I see that record 22 has a value of 0 at (0-based) index 15. And record 23 has a value of 2 at index 15.

Here's a selection from [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=grib2_#fcst) of the MET User's Guide:
```
The “GRIB2_ipdtmpl_index” and “GRIB2_ipdtmpl_val” entries are arrays of integers which specify the product description template values to be used. The indices are 0-based. For example, use the following to request a GRIB2 record whose 9-th and 27-th product description template values are 1 and 2, respectively:

GRIB2_ipdtmpl_index=[8, 26]; GRIB2_ipdtmpl_val=[1, 2];
```

I'd recommend the following...

Plot the average field and use `set_attr_name` to modify its name.
```
plot_data_plane WAFS_blended_2022110300f18.grib2 ICIP_AVG.ps 'name="ICIP"; level="P500"; GRIB2_ipdtmpl_index=15; GRIB2_ipdtmpl_val=0; set_attr_name="ICIP_AVG";' -title ICIP_AVG -plot_range 0 1 -v 3
...
DEBUG 3: MetGrib2DataFile::data_plane() -> Found exact match for 'ICIP/P500' in GRIB2 record 22 field 1 of GRIB2 file 'WAFS_blended_2022110300f18.grib2'
```
Plot the maximum field and use `set_attr_name` to modify its name.
```
plot_data_plane WAFS_blended_2022110300f18.grib2 ICIP_MAX.ps 'name="ICIP"; level="P500"; GRIB2_ipdtmpl_index=15; GRIB2_ipdtmpl_val=2; set_attr_name="ICIP_MAX";' -title ICIP_MAX -plot_range 0 1 -v 3
...
DEBUG 3: MetGrib2DataFile::data_plane() -> Found exact match for 'ICIP/P500' in GRIB2 record 23 field 1 of GRIB2 file 'WAFS_blended_2022110300f18.grib2'
```

For comparison purposes, here's what those 2 plots look like:
![Screen Shot 2022-11-07 at 9 44 34 AM](https://user-images.githubusercontent.com/21087144/200366323-458bdd71-09d3-4dc3-bc05-551e8087d7ea.png)

I think using these config options would make life easier in the long run. Please take a look at the existing *GRIB2_* filter options and please let me know if there's another one you think we should add to filter this directly instead of having to resort to `GRIB2_ipdtmpl_index` and `GRIB2_ipdtmpl_val`.

**Reply by YaliMao-NOAA** [reply]
Hi John,
Thank you for the instructions. I didn't use 'plot_data_plane' before and am not sure whether it is installed on WCOSS2.
But I did basically the same with 'wgrib2' as you did with 'plot_data_plane': separated ICIPmean and ICIPmax into two files, renamed the fields (as you did by set_attr_name), then changed the template numbers. In both your and my approaches, we need to define our own MET_GRIB_TABLES

**Reply by JohnHalleyGotway** [reply]
Yali, plot_data_plane is one of the MET utilities that's always compiled. Once you've loaded the MET modules on WCOSS, just try running `which plot_data_plane` and you should see it in your path. It's a great tool for testing what data MET is reading from input gridded datasets.

If you use the `GRIB2_ipdtmpl_index` and `GRIB2_iptdmpl_val` options, I do not see a need to define custom GRIB tables using MET_GRIB_TABLES. But of course there's any number of ways to solve the same problem.

Please let me know if you have additional questions or could use additional support.

**Reply by JohnHalleyGotway** [reply]
Just wanted to follow up on this discussion to see if questions remain. If not, please feel free to select the "Mark As Answer" button for one of the previous responses, and then I'll lock it to prevent future posts. If I don't hear otherwise in the next week or so, I'll just go ahead and mark it as being answered.

Thanks!
